### PR TITLE
Add slack hook filters and advanced formatting

### PR DESCRIPTION
**Pull request by DeniDoman**
We have faced with 3 issues in a current slack flow:

1. We want to get notification about only failed builds, not all builds.
2. We have dozens of branches, we want to know only about failed builds from **develop**, **master** and **release-\*** branches.
3. We need more information in slack message (commit message), also need to fix confusion with skipped/pending (https://github.com/cypress-io/cypress/issues/3092)

So, I have added:

1. Result filter (ALL, ONLY_FAILED, ONLY_SUCCESSFUL). Required field, default value is ALL.  
2. Branch filter (with '*' and '?' wildcards support). Optional field, empty state means filter is off, all branches are tracked. 
3. New Slack message layout.

Edit hook before:
image

Edit hook after:
image

Slack message before:
image

Slack message after:
Passed:
image
Failed:
image

### CODE REVIEW

**Comment by DeniDoman** [comment]
Please don't merge. Regex idea was not very good: branches with name like "agg23/Typescript4.2.3" doesn't work correctly in filter.

File: packages/director/src/lib/hooks/utils.ts [hunk 1]
```diff
@@ -1,24 +1,80 @@
...
+    if (!hook.hookEvents || !hook.hookEvents.length) {
+      return true;
+    }
+
+    if (hook.hookEvents.includes(event)) {
+      return true;
+    }
   }
 
   return false;
```
Comment (agoldis) [inline packages/director/src/lib/hooks/utils.ts:70]: That becomes a bit bulky to read, would you mind please extracting to a separate function, smth like `shouldReportSlackHook`

File: packages/director/src/lib/hooks/reporters/controller.ts [hunk 2]
```diff
@@ -54,6 +55,7 @@ export function reportToHook({
         reportToGenericWebHook({
           hook,
           runId: run.runId,
+          branch: run.meta.commit.branch,
```
Comment (agoldis) [inline packages/director/src/lib/hooks/reporters/controller.ts:58]: 👍🏻

File: packages/api/src/schema/schema.graphql [hunk 3]
```diff
@@ -181,6 +189,8 @@ input HookInput {
   bitbucketUsername: String
   bitbucketToken: String
   bitbucketBuildName: String
+  slackEventFilter: String
+  slackBranchFilter: [String]
```
Comment (agoldis) [inline packages/api/src/schema/schema.graphql:193]: That allows `[null]`, which is not something we want I guess :) 

How about `[String!]`?

File: packages/api/src/schema/schema.graphql [hunk 4]
```diff
@@ -162,6 +168,8 @@ type Hook {
...
```
Comment (agoldis) [inline packages/api/src/schema/schema.graphql:172]: That's very unfortunate duplication :( Would be nice to eliminate it eventually

**Review by agoldis — COMMENTED** [review]
Very nice @DeniDoman! Thanks a lot for this PR!

 I have added a few minor comments.

File: packages/api/src/schema/schema.graphql (hunk 3 above)
Comment (DeniDoman) [reply packages/api/src/schema/schema.graphql:193]: I see that the whole Hook type doesn't contain required fields. I thought this some tricky plan, so I leave new fields not required too. 

```
type Hook {
  hookId: String
  url: String
  headers: String
  hookEvents: [String]
  hookType: String
  githubToken: String
  githubContext: String
  bitbucketUsername: String
  bitbucketToken: String
  bitbucketBuildName: String
  slackEventFilter: String
  slackBranchFilter: [String]
}
```

Should I add a required flag only to my fields or for all relevant fields? BTW branch filter can be empty or null, but slackEventFilter must have a value.

File: packages/director/src/lib/hooks/reporters/slack.ts [hunk 5]
```diff
@@ -1,46 +1,67 @@
...
+    } *Skipped:* ${pending}\n\n\n` +
+    `${failures + skipped > 0 ? ':red_circle:' : ':white_circle:'} *Failed*: ${
+      failures + skipped
+    }`;
+
+  const commitDescription = `*Branch:*\n${branch}\n\n*Commit:*\n${
+    message.length > 100 ? `${message.substring(0, 100)}...` : message
+  }`;
 
   if (hookEvent === HookEvent.RUN_FINISH) {
```
Comment (DeniDoman) [inline packages/director/src/lib/hooks/reporters/slack.ts:64]: @agoldis could you please explain why we need this 5 seconds timeout? Can't get it.

Comment (DeniDoman) [reply packages/director/src/lib/hooks/reporters/slack.ts:64]:
```
  if (hookEvent === HookEvent.RUN_FINISH) {
    await new Promise((resolve) => {
      setTimeout(() => {
        resolve(null);
      }, 5000);
    });
  }
```

File: packages/api/src/schema/schema.graphql (hunk 3 above)
Comment (DeniDoman) [reply packages/api/src/schema/schema.graphql:193]: Also empty or null slackBranchFilter means all branches are tracked. So it's a valid state.

File: packages/director/src/lib/hooks/reporters/slack.ts (hunk 5 above)
Comment (DeniDoman) [reply packages/director/src/lib/hooks/reporters/slack.ts]: I have removed it so far. Say if need to revert.

File: packages/api/src/schema/schema.graphql (hunk 4 above)
Comment (DeniDoman) [reply packages/api/src/schema/schema.graphql]: Agree, but seems as out of scope of the change.

File: packages/api/src/schema/schema.graphql (hunk 3 above)
Comment (DeniDoman) [reply packages/api/src/schema/schema.graphql:193]: I have left it as is.

File: packages/director/src/lib/hooks/utils.ts (hunk 1 above)
Comment (DeniDoman) [reply packages/director/src/lib/hooks/utils.ts]: Totally agree, thank you. I have split it into 4 functions. Also I have covered all of them by unit tests 😎

File: packages/director/src/lib/hooks/reporters/slack.ts (hunk 5 above)
Comment (agoldis) [reply packages/director/src/lib/hooks/reporters/slack.ts]: Doesn't seem like those are required.

**Review by agoldis — APPROVED** [review]

**Pull request merged**